build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -0,0 +1,158 @@
+          containsArgumentsUsage = true;
+          path.stop();
+        }
+      },
+    });
+    return containsArgumentsUsage;
+  },
+
+  // If the function contains a ThisExpression, converting from a FunctionDeclaration to a
+  // VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+  // Bail out on modification.
+  containsThisExpression: (t, path) => {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  },
+
+  // If the FunctionDeclaration identifier is newed in the scope of this program, bail out on modification.
+  isNewedInProgramScope: (t, path) => {
+    const {name} = path.get('id').node;
+    let isNewed = false;
+    path
+      .findParent(path => path.isProgram())
+      .traverse({
+        NewExpression(path) {
+          if (t.isIdentifier(path.node.callee, {name})) {
+            isNewed = true;
+            path.stop();
+          }
+        },
+      });
+    return isNewed;
+  },
+};
+
+function createVariableDeclaration(t, path) {
+  const {params, body, async, id} = t.cloneNode(path.node);
+  const arrowFunction = t.arrowFunctionExpression(
+    params,
+    body.body[0].argument,
+    async
+  );
+  const declarations = t.variableDeclarator(
+    t.identifier(id.name),
+    arrowFunction
+  );
+
+  // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
+  // The name was already reserved for the FunctionDeclaration.
+  return t.variableDeclaration('let', [declarations]);
+}
+
+// Attempt to convert simple single ReturnStatement FunctionDeclarations to ArrowFunctionExpressions.
+// See BAIL_OUT_CONDITIONS for reasons why FunctionDeclarations would not be modified.
+module.exports = function({types: t}) {
+  const DEBUG = false;
+  const REPORT = false;
+
+  return {
+    name: 'safe-arrows',
+    pre() {
+      this.bailoutCount = {...BAIL_OUT_CONDITIONS};
+      Object.keys(this.bailoutCount).forEach(
+        key => (this.bailoutCount[key] = [])
+      );
+      this.successCount = [];
+    },
+    visitor: {
+      FunctionDeclaration(path) {
+        const name = (path.get('id') && path.get('id').node.name) || '';
+        for (const [id, method] of Object.entries(BAIL_OUT_CONDITIONS)) {
+          if (method(t, path)) {
+            this.bailoutCount[id].push(name);
+            if (DEBUG) {
+              console /*OK*/
+                .log(
+                  `Bail on ${
+                    name !== '' ? `function ${name}:` : 'item:'
+                  } Reason ${id}.`
+                );
+            }
+            return;
+          }
+        }
+        this.successCount.push(name);
+        if (DEBUG) {
+          console /*OK*/
+            .log(`Success for function ${path.get('id').node.name}.`);
+        }
+        path.replaceWith(createVariableDeclaration(t, path));
```

Comment:
This is not 100% safe. `FunctionDeclaration`s are available to be used at _any_ time in their enclosing scope, even before the program would reach the line that define the function. Replacing it with a variable declaration does not preserve that.

The only way to make this safe is to find the enclosing block, and insert this variable as the first parameter of the body.